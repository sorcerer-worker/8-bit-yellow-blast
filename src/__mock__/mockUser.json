{
    "characterName": "Test User",
    "discordId": "757938182483485954",
    "class": "warrior",
    "pouch": 98638,
    "pvp": false,
    "inventory": {
      "weapons": [],
      "consumables": [],
      "classSkills": [],
      "otherSkills": []
    },
    "health": {
      "currentHealth": 10,
      "maxHealth": 10
    },
    "level": {
      "currentLevel": 1
    },
    "stats": {
      "strength": 5,
      "defense": 5,
      "speed": 28,
      "vitality": 10,
      "intelligence": 5,
      "magicResistance": 5
    }
  }