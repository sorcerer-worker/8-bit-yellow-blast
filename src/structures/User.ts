import { ActionManager } from "../managers/ActionManager";

/**
 * Represents the `UserData` type used by `User``.
 */
export type UserData = {
    [key: string]: any,
    name: string,
    id: number
};

export interface ForceUserPrevent {
    /**
     * Prevents forms of overload.
     */
    overload: {
        /**
         * Represents Health.
         */
        hp: boolean
    }
}

/**
 * Represents a `User` using the `BattleManager`.
 */
export class User {

    public data: UserData;
    public actions: ActionManager;
    public prevent: ForceUserPrevent = { // <-- TODO: It could use additional prevention options
        overload: {
            hp: true
        }
    };

    /**
     * Builds a new `User` using `UserData`.
     * @param {UserData} data  Object based data about the `User`
     */
    constructor(data: UserData) {
        this.data = data;
        this.actions = new ActionManager([]); // <-- Planning to pass this information through this.data
    }
    /**
     * Returns the `User`'s name property.
     */
    public get name(): string{
        return this.data.name;
    }
    /**
     * Returns the `User`'s currentHealth.
     */
    public get currentHealth(): number {
        return this.data.health.currentHealth
    }
    public set currentHealth(amount: number) {
        this.data.health.currentHealth = amount
        if (this.prevent.overload.hp) {
            if (this.currentHealth > this.data.health.maxHealth) {
                this.currentHealth = this.data.health.maxHealth;
            }
            if (this.currentHealth < 0) {
                this.currentHealth = 0
            }
        }
    }
}