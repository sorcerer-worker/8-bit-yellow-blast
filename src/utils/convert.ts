import { User, UserData } from "../structures/User";

type Configuration = {
    name: string,
    id: string
}

function convert(user: any, config: Configuration): UserData {
    let newUserData: UserData = {
        name: config.name ? user[config.name] : user["name"],
        id: config.id ? user[config.id] : user["id"],
        ...user // Add remaining info
    }
    if (config.name) {
        // Deletes duplicate
        delete newUserData[config.name]
    }
    if (config.id) {
        // Deletes duplicate
        delete newUserData[config.id]
    }
    return newUserData;
}

export { Configuration, convert };