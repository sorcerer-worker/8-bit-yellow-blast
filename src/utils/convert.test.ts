import { UserData } from "../structures/User";
import { convert } from "./convert";

// json files
import mockUser from "../__mock__/mockUser.json"
import mockConfig from "../__mock__/mockConfig.json"

let converted: UserData;

describe("convert utility tool", () => {
    beforeAll(() => {
        converted = convert(mockUser, mockConfig)
    })
    it("runs successful conversion", () => {
        expect(converted).toMatchSnapshot();
    })
})