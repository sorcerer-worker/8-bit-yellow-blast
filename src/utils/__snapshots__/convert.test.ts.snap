// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert utility tool runs successful conversion 1`] = `
Object {
  "class": "warrior",
  "health": Object {
    "currentHealth": 10,
    "maxHealth": 10,
  },
  "id": 757938182483486000,
  "inventory": Object {
    "classSkills": Array [],
    "consumables": Array [],
    "otherSkills": Array [],
    "weapons": Array [],
  },
  "level": Object {
    "currentLevel": 1,
  },
  "name": "Test User",
  "pouch": 98638,
  "pvp": false,
  "stats": Object {
    "defense": 5,
    "intelligence": 5,
    "magicResistance": 5,
    "speed": 28,
    "strength": 5,
    "vitality": 10,
  },
}
`;
